File: src/add-options.ts
import { scanArgs } from './args.ts'
import { ADD_PARSE_OPTIONS } from './commands.ts'
import { type TaskComplexity, TaskComplexitySchema } from './task.ts'

const ADD_USAGE =
  'usage: factory add [--raw] [--trivial | --complexity trivial|complex] [--allow-dirty] ' +
  '[--name <slug>] [intent...] [--verify <cmd...>] [--edit]'

export type ParsedAddOptions = {
  args: string[]
  raw: boolean
  complexity: TaskComplexity | null
  // Start a new task on a worktree that already has uncommitted changes. The
  // task's commit stages the whole tree (git add -A), so preexisting edits
  // would be swept into it under a message that describes only the new task —
  // starting dirty is an error unless it is deliberate.
  allowDirty: boolean
  // Explicit task id/slug from the caller (spawner tools already named the
  // worktree) — skips the namer model call entirely.
  name: string | null
}

export type ParseAddOptionsResult =
  | { ok: true; options: ParsedAddOptions }
  | { ok: false; message: string }

function fail(message: string): ParseAddOptionsResult {
  return { ok: false, message: `${ADD_USAGE}\n${message}` }
}

function setComplexity(
  current: TaskComplexity | null,
  next: TaskComplexity
): ParseAddOptionsResult | TaskComplexity {
  if (current !== null && current !== next) {
    return fail(`conflicting complexity flags: ${current} and ${next}`)
  }
  return next
}

export function parseAddOptions(args: string[]): ParseAddOptionsResult {
  // Unknown flag-ish tokens are intent words ('collect'); --verify consumes the
  // rest of the line as its tail. --edit is intentionally undeclared here so it
  // stays in the intent tokens for resolveIntent (which strips it anywhere,
  // including inside the verify tail).
  const scan = scanArgs(ADD_PARSE_OPTIONS, args, { unknown: 'collect' })
  if (!scan.ok) {
    if (scan.error.option === '--name') {
      return fail('--name needs a value')
    }
    return fail('--complexity needs a value: trivial or complex')
  }

  const tail = scan.flags['--verify']
  if (tail?.includes('--trivial') || tail?.includes('--complexity')) {
    return fail('complexity flags must appear before --verify')
  }

  let name: string | null = null
  for (const value of scan.flags['--name']) {
    if (value.startsWith('--')) {
      return fail('--name needs a value')
    }
    name = value
  }

  let complexity: TaskComplexity | null = null
  if (scan.flags['--trivial']) {
    complexity = 'trivial'
  }
  for (const value of scan.flags['--complexity']) {
    if (value.startsWith('--')) {
      return fail('--complexity needs a value: trivial or complex')
    }
    const parsed = TaskComplexitySchema.safeParse(value)
    if (!parsed.success) {
      return fail(`invalid complexity "${value}" (expected trivial or complex)`)
    }
    const next = setComplexity(complexity, parsed.data)
    if (typeof next !== 'string') {
      return next
    }
    complexity = next
  }

  return {
    ok: true,
    options: {
      args: tail === undefined ? scan.positionals : [...scan.positionals, '--verify', ...tail],
      raw: scan.flags['--raw'],
      complexity,
      allowDirty: scan.flags['--allow-dirty'],
      name,
    },
  }
}

File: src/add_options.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TaskComplexity {
    Trivial,
    Complex,
}

impl TaskComplexity {
    fn parse(value: &str) -> Option<Self> {
        match value {
            "trivial" => Some(Self::Trivial),
            "complex" => Some(Self::Complex),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Trivial => "trivial",
            Self::Complex => "complex",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedAddOptions {
    pub args: Vec<String>,
    pub raw: bool,
    pub complexity: Option<TaskComplexity>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ParseAddOptionsResult {
    Ok(ParsedAddOptions),
    Err(String),
}

const ADD_USAGE: &str =
    "usage: factory add [--raw] [--trivial | --complexity trivial|complex] [intent...] [--verify <cmd...>] [--edit]";

fn fail(message: impl AsRef<str>) -> ParseAddOptionsResult {
    ParseAddOptionsResult::Err(format!("{ADD_USAGE}\n{}", message.as_ref()))
}

fn set_complexity(
    current: Option<TaskComplexity>,
    next: TaskComplexity,
) -> Result<TaskComplexity, String> {
    if let Some(current) = current {
        if current != next {
            return Err(format!(
                "conflicting complexity flags: {} and {}",
                current.as_str(),
                next.as_str()
            ));
        }
    }
    Ok(next)
}

pub fn parse_add_options(args: &[String]) -> ParseAddOptionsResult {
    let verify_index = args.iter().position(|arg| arg == "--verify");
    let (head, tail): (&[String], &[String]) = match verify_index {
        Some(i) => (&args[..i], &args[i + 1..]),
        None => (args, &[]),
    };
    if tail
        .iter()
        .any(|arg| arg == "--trivial" || arg == "--complexity")
    {
        return fail("complexity flags must appear before --verify");
    }

    let mut cleaned = Vec::new();
    let mut raw = false;
    let mut complexity = None;
    let mut i = 0;
    while i < head.len() {
        let arg = &head[i];
        if arg == "--raw" {
            raw = true;
            i += 1;
            continue;
        }
        if arg == "--trivial" {
            match set_complexity(complexity, TaskComplexity::Trivial) {
                Ok(next) => complexity = Some(next),
                Err(message) => return fail(message),
            }
            i += 1;
            continue;
        }
        if arg == "--complexity" {
            let Some(value) = head.get(i + 1) else {
                return fail("--complexity needs a value: trivial or complex");
            };
            if value.starts_with("--") {
                return fail("--complexity needs a value: trivial or complex");
            }
            let Some(parsed) = TaskComplexity::parse(value) else {
                return fail(format!(
                    "invalid complexity \"{value}\" (expected trivial or complex)"
                ));
            };
            match set_complexity(complexity, parsed) {
                Ok(next) => complexity = Some(next),
                Err(message) => return fail(message),
            }
            i += 2;
            continue;
        }
        cleaned.push(arg.clone());
        i += 1;
    }

    let args = if verify_index.is_some() {
        let mut full = cleaned;
        full.push("--verify".to_string());
        full.extend(tail.iter().cloned());
        full
    } else {
        cleaned
    };

    ParseAddOptionsResult::Ok(ParsedAddOptions {
        args,
        raw,
        complexity,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn strings(args: &[&str]) -> Vec<String> {
        args.iter().map(|arg| (*arg).to_string()).collect()
    }

    fn expect_parsed(args: &[&str]) -> ParsedAddOptions {
        match parse_add_options(&strings(args)) {
            ParseAddOptionsResult::Ok(options) => options,
            ParseAddOptionsResult::Err(message) => panic!("{message}"),
        }
    }

    fn expect_error(args: &[&str]) -> String {
        match parse_add_options(&strings(args)) {
            ParseAddOptionsResult::Ok(_) => panic!("expected parse failure"),
            ParseAddOptionsResult::Err(message) => message,
        }
    }

    #[test]
    fn parses_trivial() {
        assert_eq!(
            expect_parsed(&["--trivial", "Fix", "typo"]),
            ParsedAddOptions {
                args: strings(&["Fix", "typo"]),
                raw: false,
                complexity: Some(TaskComplexity::Trivial),
            }
        );
    }

    #[test]
    fn parses_complexity_complex() {
        assert_eq!(
            expect_parsed(&["--complexity", "complex", "Refactor", "parser"]),
            ParsedAddOptions {
                args: strings(&["Refactor", "parser"]),
                raw: false,
                complexity: Some(TaskComplexity::Complex),
            }
        );
    }

    #[test]
    fn accepts_redundant_trivial_declarations() {
        assert_eq!(
            expect_parsed(&["--trivial", "--complexity", "trivial", "Fix", "typo"]).complexity,
            Some(TaskComplexity::Trivial)
        );
    }

    #[test]
    fn rejects_conflicting_complexity_declarations() {
        assert!(
            expect_error(&["--trivial", "--complexity", "complex", "Fix"])
                .contains("conflicting complexity flags")
        );
    }

    #[test]
    fn rejects_missing_complexity_value() {
        assert!(expect_error(&["--complexity"])
            .contains("--complexity needs a value: trivial or complex"));
    }

    #[test]
    fn rejects_invalid_complexity_value() {
        assert!(expect_error(&["--complexity", "maybe"])
            .contains("invalid complexity \"maybe\" (expected trivial or complex)"));
    }

    #[test]
    fn parses_complexity_before_verify() {
        assert_eq!(
            expect_parsed(&[
                "--complexity",
                "trivial",
                "Fix",
                "typo",
                "--verify",
                "bun",
                "test"
            ]),
            ParsedAddOptions {
                args: strings(&["Fix", "typo", "--verify", "bun", "test"]),
                raw: false,
                complexity: Some(TaskComplexity::Trivial),
            }
        );
    }

    #[test]
    fn rejects_complexity_flags_after_verify() {
        assert!(
            expect_error(&["Fix", "--verify", "bun", "test", "--trivial"])
                .contains("complexity flags must appear before --verify")
        );
        assert!(
            expect_error(&["Fix", "--verify", "bun", "test", "--complexity"])
                .contains("complexity flags must appear before --verify")
        );
    }

    #[test]
    fn preserves_normal_verify_tokens_after_verify() {
        assert_eq!(
            expect_parsed(&["Fix", "typo", "--verify", "bun", "test", "--raw"]),
            ParsedAddOptions {
                args: strings(&["Fix", "typo", "--verify", "bun", "test", "--raw"]),
                raw: false,
                complexity: None,
            }
        );
    }

    #[test]
    fn parses_raw_before_verify() {
        assert_eq!(
            expect_parsed(&["--raw", "Fix", "typo"]),
            ParsedAddOptions {
                args: strings(&["Fix", "typo"]),
                raw: true,
                complexity: None,
            }
        );
    }
}
